server/src/instant/db/transaction.clj

```diff
@@ -181,39 +269,50 @@
                                     :num-tx-steps (count tx-steps)
                                     :detailed-tx-steps (pr-str tx-steps)}}
      (prevent-system-catalog-updates! app-id opts)
-     (let [results
+     (let [tx-steps (if (some #(= :delete-entity (first %)) tx-steps)
+                      (->> tx-steps
+                           (resolve-lookups-for-delete-entity conn app-id)
+                           (resolve-etypes-for-delete-entity conn app-id)
+                           (expand-delete-entity-cascade conn app-id))
+                      tx-steps)
+
+           results
            (reduce
-            (fn [acc [op & args]]
+            (fn [acc [op args]]
               (when (#{:add-attr :update-attr} op)
                 (prevent-system-catalog-attrs-updates! op args))
-              (let [res (case op
-                          :add-attr
-                          (attr-model/insert-multi! conn app-id args)
-                          :delete-attr
-                          (attr-model/delete-multi! conn app-id args)
-                          :update-attr
-                          (attr-model/update-multi! conn app-id args)
-                          :delete-entity
-                          (triple-model/delete-entity-multi! conn app-id args)
-                          :add-triple
-                          (triple-model/insert-multi! conn attrs app-id args)
-                          :deep-merge-triple
-                          (triple-model/deep-merge-multi! conn attrs app-id args)
-                          :retract-triple
-                          (triple-model/delete-multi! conn app-id args))]
-                (assoc acc op res)))
+              (case op
+                :add-attr
+                (assoc acc :add-attr (attr-model/insert-multi! conn app-id args))
+
+                :delete-attr
+                (assoc acc :delete-attr (attr-model/delete-multi! conn app-id args))
+
+                :update-attr
+                (assoc acc :update-attr (attr-model/update-multi! conn app-id args))
+
+                :delete-entity-no-cascade
```

Comment:
Why have `:delete-entity-no-cascade` as an operation? Do we expect users to use it or is it just for our internal usage?

Reply:
The idea here is that we don’t try to expand `:delete-entity` second time, since some methods call `transact-without-tx-conn!` directly and some `permissioned-transaction/transact!`